### PR TITLE
Add StagingProject model

### PR DESCRIPTION
**Pull request by bgeuken**

### CODE REVIEW

File: src/api/app/models/staging_project.rb [hunk 1]
```diff
@@ -97,47 +97,46 @@ def check_state
...
         end
+
         @building_repositories << current_repo
       end
     end
-    if @building_repositories.present?
-      @broken_packages = @broken_packages.reject { |p| p['state'] == 'unresolvable' }
-    end
+
+    @broken_packages.reject! { |p| p['state'] == 'unresolvable' } if @building_repositories.present?
```
Comment (mdeniz) [inline src/api/app/models/staging_project.rb]: Better not to still use `p` as a variable name

File: src/api/app/models/project.rb [hunk 2]
```diff
@@ -78,7 +78,6 @@ def autocomplete(search)
   has_many :target_of_bs_requests, through: :target_of_bs_request_actions, source: :bs_request
 
   has_one :staging, class_name: 'StagingWorkflow', inverse_of: :project
-  belongs_to :staging_workflow, inverse_of: :staging_projects
   has_many :staged_requests, class_name: 'BsRequest', foreign_key: 'staging_project_id', dependent: :nullify
```
Comment (DavidKang) [inline src/api/app/models/project.rb]: I think this association is not needed anymore because it was moved to `staging_project`. BTW, `dependent: :nullify` is not included in the assocation defined in [staging_project](https://github.com/openSUSE/open-build-service/pull/6224/files#diff-61726d0667701e350f427c3907255462R3)

Comment (bgeuken) [reply src/api/app/models/project.rb]: I had to re-add this because it's used in some existing code. Will check how this can be resolved

File: src/api/app/models/staging_project.rb [hunk 3]
```diff
@@ -0,0 +1,142 @@
+# The staging project is the place where we collect a selection of packages and test, build and review them.
+class StagingProject < Project
```
Comment (ChrisBr) [inline src/api/app/models/staging_project.rb]: [nitpick] I'm still in favor of introducing a global namespace ``Staging`` and put everything staging workflow related into it:

```
Staging::Project
Staging::Workflow
Staging::RequestsController
Staging::IgnoredRequestsController
etc...
```

File: src/api/app/models/staging_project.rb [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  has_many :staged_requests, class_name: 'BsRequest', foreign_key: :staging_project_id, dependent: :nullify
+  has_many :status_reports, through: :repositories
+  belongs_to :staging_workflow, inverse_of: :staging_projects
+
+  def staging_identifier
+    name[/.*:Staging:(.*)/, 1]
+  end
+
+  def untracked_requests
+    requests_to_review - staged_requests
```
Comment (ChrisBr) [inline src/api/app/models/staging_project.rb]: I would really like to prefer to use proper ActiveRecord collections instead off doing these array calculations. In https://github.com/openSUSE/open-build-service/pull/6190 @vpereira and me actually needed to refactor some of this stuff introduced earlier because it was just to difficult to work with arrays instead of an ActiveRecord collection ...

I think a simple solution like this could even work:

```
requests_to_review.where.not(staged_requests)
```

File: src/api/app/models/staging_project.rb [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    attribs = [:by_group, :by_user, :by_project, :by_package]
+
+    (requests_to_review + staged_requests).uniq.each do |request|
+      request.reviews.each do |review|
+        next if review.state.to_s == 'accepted' || review.by_project == name
+        # FIXME: this loop (and the inner if) would not be needed
+        # if every review only has one valid by_xxx.
+        # I'm keeping it to mimic the python implementation.
+        # Instead, we could have something like
+        # who = review.by_group || review.by_user || review.by_project || review.by_package
```
Comment (ChrisBr) [inline src/api/app/models/staging_project.rb]: I guess this is c&p right? What is the problem with the or? 

cc @coolo

File: src/api/app/controllers/staging_project_controller.rb [hunk 6]
```diff
@@ -4,7 +4,7 @@ class StagingProjectController < ApplicationController
   private
 
   def set_project
-    @project = Project.find_by(name: params[:project])
+    @project = StagingProject.find_by(name: params[:project])
```
Comment (dmarcoux) [inline src/api/app/controllers/staging_project_controller.rb]: I just realized now, while we are here, let's use `find_by!`.

File: src/api/app/models/staging_project.rb (hunk 5 above)
Comment (dmarcoux) [reply src/api/app/models/staging_project.rb]: @ChrisBr Yes this is a copy and paste.

File: src/api/app/models/staging_project.rb (hunk 4 above)
Comment (coolo) [reply src/api/app/models/staging_project.rb]: 'I would really like to prefer'? How polite you are :)

File: src/api/app/models/staging_project.rb (hunk 5 above)
Comment (dmarcoux) [reply src/api/app/models/staging_project.rb]: Since a review's `by_user`, `by_group`, `by_project` and `by_package` can all be defined (but they don't have to), we need to go through all of them. With the `or`, we would only check the first one which is defined (even though there is the possibility of having another `by_...` defined afterwards that`.

File: src/api/app/models/staging_project.rb [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    set_buildinfo
+
+    return :building if building_repositories.present?
+    return :failed if broken_packages.present?
+
+    :acceptable
+  end
+
+  def relevant_status_reports
+    @relevant_status_reports ||= status_reports.where(uuid: repositories.map(&:build_id))
```
Comment (Ana06) [inline src/api/app/models/staging_project.rb]: `map(&:build_id)`! :see_no_evil: use `pluck(:build_id)` instead!

File: src/api/app/models/staging_project.rb [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+  end
+
+  def missing_checks?
+    relevant_status_reports.any? { |report| report.missing_checks.present? }
+  end
+
+  def check_state
+    status_checks = Status::Check.where(status_reports_id: relevant_status_reports)
```
Comment (Ana06) [inline src/api/app/models/staging_project.rb]: `status_reports` instead `status_reports_id`

File: src/api/app/models/staging_project.rb [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    return :testing if missing_checks? || status_checks.pending.exists?
+    return :failed if status_checks.failed.exists?
+    return :acceptable
+  end
+
+  def set_buildinfo
+    buildresult = Buildresult.find_hashed(project: name, view: 'summary')
+
+    # data of each pkg/repo/arch combo with state broken or failed
```
Comment (Ana06) [inline src/api/app/models/staging_project.rb]: I don't think this comment is needed, code is easier to understand than the comment. If you think it is not clear, I would prefer to change the variable name than to keep the comment.

File: src/api/app/models/staging_project.rb [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+      result.elements('status') do |status|
+        code = status.get('code')
+
+        if code.in?(['broken', 'failed']) || (code == 'unresolvable' && !building)
+          @broken_packages << { package:    status['package'],
+                                project:    name,
+                                state:      code,
+                                details:    status['details'],
+                                repository: result['repository'],
+                                arch:       result['arch'] }.with_indifferent_access
```
Comment (Ana06) [inline src/api/app/models/staging_project.rb]: I would drop the `with_indifferent_access`. I don't see the advantage of being able to access with a string.

File: src/api/app/models/staging_project.rb (hunk 3 above)
Comment (bgeuken) [reply src/api/app/models/staging_project.rb]: Hm, not sure if that couldn't collide in some way.

In the end we would have to write something like, unless we rename it:

```ruby
module Staging
  class Project < ::Project
```

File: src/api/app/models/staging_project.rb [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    @broken_packages = []
+    # data and count of all repos that are currently building
+    @building_repositories = []
+
+    buildresult.elements('result') do |result|
+      building = ['published', 'unpublished'].exclude?(result['state']) || result['dirty'] == 'true'
+
+      result.elements('status') do |status|
+        code = status.get('code')
+
```
Comment (Ana06) [inline src/api/app/models/staging_project.rb]: [nitpick] I would remove this empty line, there are not conceptually different code blocks

File: src/api/app/models/staging_project.rb (hunk 4 above)
Comment (ChrisBr) [reply src/api/app/models/staging_project.rb]: @coolo thank my parents to raise such a polity young boy :smile: :family_man_woman_boy:

Comment (bgeuken) [reply src/api/app/models/staging_project.rb]: `requests_to_review` currently returns an Array. We have some follow up commits that replaces the existing code with scopes. Once this is merged I would open another PR and update the `untracked_requests` method as well.

File: src/api/app/models/staging_project.rb (hunk 3 above)
Comment (ChrisBr) [reply src/api/app/models/staging_project.rb]: @bgeuken I don't think this would be a problem?

File: src/api/app/models/staging_project.rb (hunk 1 above)
Comment (Ana06) [reply src/api/app/models/staging_project.rb]: I am surprised one single comment was enough... :unamused:

File: src/api/app/controllers/staging_project_controller.rb (hunk 6 above)
Comment (Ana06) [reply src/api/app/controllers/staging_project_controller.rb]: :see_no_evil:

File: src/api/app/models/staging_project.rb (hunk 3 above)
Comment (bgeuken) [reply src/api/app/models/staging_project.rb]: We named it

```ruby
module Staging
  class StagingProject < Project
```
now.

File: src/api/app/models/staging_project.rb (hunk 4 above)
Comment (bgeuken) [reply src/api/app/models/staging_project.rb]: > ChrisBr 21 minutes ago  Member
> @coolo thank my parents to raise such a polity young boy 

The only task left for them is to make you eat your vegetables:laughing:

File: src/api/app/models/staging/staging_project.rb [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+      @missing_reviews = []
+      attribs = [:by_group, :by_user, :by_project, :by_package]
+
+      (requests_to_review + staged_requests).uniq.each do |request|
+        request.reviews.each do |review|
+          next if review.state.to_s == 'accepted' || review.by_project == name
+          # FIXME: this loop (and the inner if) would not be needed
+          # if every review only has one valid by_xxx.
+          # I'm keeping it to mimic the python implementation.
```
Comment (coolo) [inline src/api/app/models/staging/staging_project.rb]: there is already Review.assigned_reviewer which does a very similiar loop

**Comment by coolo** [comment]
looking at the check code - you really want to take the coverage I added I added in #6142

File: src/api/app/models/staging_project.rb (hunk 4 above)
Comment (ChrisBr) [reply src/api/app/models/staging_project.rb]: > requests_to_review currently returns an Array. We have some follow up commits that replaces the existing code with scopes. Once this is merged I would open another PR and update the untracked_requests method as well.

Ok but we should do this ASAP as we really run into some problems in the mentioned PR...

File: src/api/app/models/staging_project.rb [hunk 13]
```diff
@@ -0,0 +1,142 @@
+# The staging project is the place where we collect a selection of packages and test, build and review them.
+class StagingProject < Project
+  has_many :staged_requests, class_name: 'BsRequest', foreign_key: :staging_project_id, dependent: :nullify
+  has_many :status_reports, through: :repositories
+  belongs_to :staging_workflow, inverse_of: :staging_projects
```
[the remark this answers is not in the document]
Comment (bgeuken) [reply src/api/app/models/staging_project.rb]: It's for having [bi-directional associations](https://guides.rubyonrails.org/association_basics.html#bi-directional-associations). It has a couple of advantages, stated in the rails guide, and as it seems there wouldn't be any drawbacks:smile:

File: src/api/app/models/staging_project.rb [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+  def missing_reviews
+    return @missing_reviews if @missing_reviews
+
+    @missing_reviews = []
+    attribs = [:by_group, :by_user, :by_project, :by_package]
+
+    (requests_to_review + staged_requests).uniq.each do |request|
+      request.reviews.each do |review|
+        next if review.state.to_s == 'accepted' || review.by_project == name
```
[the remark this answers is not in the document]
Comment (bgeuken) [reply src/api/app/models/staging_project.rb]: `review.state.to_sym == :accepted` then^^

File: src/api/app/models/staging_project.rb (hunk 5 above)
Comment (coolo) [reply src/api/app/models/staging_project.rb]: from review.rb:
```
    if validate_reviewer_fields
      errors.add(:base, 'it is not allowed to have more than one reviewer entity: by_user, by_group, by_project, by_package')
    end

File: src/api/app/models/staging_project.rb (hunk 7 above)
Comment (bgeuken) [reply src/api/app/models/staging_project.rb]: Nah, build_id is a [method](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/repository.rb#L264) that makes a backend call.

File: src/api/app/models/staging_project.rb (hunk 8 above)
Comment (bgeuken) [reply src/api/app/models/staging_project.rb]: You would get a `Mysql2::Error: Unknown column 'status_checks.status_reports'` then

File: src/api/app/models/staging/staging_project.rb (hunk 12 above)
Comment (bgeuken) [reply src/api/app/models/staging/staging_project.rb]: Oh, right. But there are a few differences. I would keep the code as is and do that in a later refactoring. We are also going to drop the `requests_to_review` as discussed yesterday.

**Comment by bgeuken** [comment]
@Ana06 @ChrisBr @coolo updated. Please have another look.

File: src/api/app/models/staging_project.rb (hunk 5 above)
Comment (coolo) [reply src/api/app/models/staging_project.rb]: if by_package is set, by_project is the project of it - there is basically no way to have by_package *without* by_project.

But by_user and by_group are mutually exclusive and 
```
MariaDB [obs_production]> select distinct number,by_package,by_project from reviews v join bs_requests r on r.id=v.bs_request_id where by_user is not null and by_group is not null;
+--------+------------+------------+
| number | by_package | by_project |
+--------+------------+------------+
| 491574 | NULL       | NULL       |
+--------+------------+------------+
```

And https://build.opensuse.org/request/show/491574 looks like an exercise In free love (if you don't know the reference, you're lost on music :)

File: src/api/app/models/staging_project.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/models/staging_project.rb]: you can have a bi-directional association without that :confused: What are the advantages? :confused:

File: src/api/app/models/staging_project.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/models/staging_project.rb]: @bgeuken 

> `review.state.to_sym == :accepted` then^^

why? we have a validation in `Review` that ensures that the sate is one of `:new, :declined, :accepted, :superseded, :obsoleted`. All of them are symbols, so `review.state == :accepted`

File: src/api/app/models/staging_project.rb (hunk 7 above)
Comment (Ana06) [reply src/api/app/models/staging_project.rb]: :scream: ok

File: src/api/app/models/staging_project.rb [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    @broken_packages
+  end
+
+  def missing_reviews
+    return @missing_reviews if @missing_reviews
+
+    @missing_reviews = []
+    attribs = [:by_group, :by_user, :by_project, :by_package]
+
+    (requests_to_review + staged_requests).uniq.each do |request|
```
[the remark this answers is not in the document]
Comment (bgeuken) [reply src/api/app/models/staging_project.rb]: Forgot to mention it here. The requests_to_review will soon die. Though I can change the code for now.

File: src/api/app/models/staging_project.rb (hunk 8 above)
Comment (Ana06) [reply src/api/app/models/staging_project.rb]: It is wrong defined here: https://github.com/openSUSE/open-build-service/pull/6224/files#diff-8d266b19374f8906dc972ad3d776f346R4

File: src/api/app/models/staging_project.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (dmarcoux) [reply src/api/app/models/staging_project.rb]: It makes the bi-directional associations explicit and they don't revolve on the Rails magic to work. This magic doesn't work when the options `:through` or `:foreign_key` are used. It's all listed in the link that @bgeuken shared in his comment.

File: src/api/app/models/staging_project.rb (hunk 8 above)
Comment (bgeuken) [reply src/api/app/models/staging_project.rb]: Might be, but I don't find a fix for this right now... Having a look later.

File: src/api/app/models/status/check.rb [hunk 16]
```diff
@@ -23,6 +23,8 @@ class Status::Check < ApplicationRecord
   #### Callbacks macros: before_save, after_save, etc.
 
   #### Scopes (first the default_scope macro if is used)
+  scope :pending, -> { where(state: 'pending') }
+  scope :failed, -> { where(state: 'failure') }
```
Comment (coolo) [inline src/api/app/models/status/check.rb]: what about checks that are state=error?

**Comment by bgeuken** [comment]
@Ana06 Updated. Please check again

File: src/api/app/models/staging_project.rb [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+      request.reviews.each do |review|
+        next if review.state.to_s == 'accepted' || review.by_project == name
+        # FIXME: this loop (and the inner if) would not be needed
+        # if every review only has one valid by_xxx.
+        # I'm keeping it to mimic the python implementation.
```
[the remark this answers is not in the document]
Comment (bgeuken) [reply src/api/app/models/staging_project.rb]: That method is just copied over from the old staging project code. We kept it for reference.

File: src/api/app/models/staging/staging_project.rb [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+
+    def relevant_status_reports
+      @relevant_status_reports ||= status_reports.where(uuid: repositories.map(&:build_id))
+    end
+
+    def missing_checks?
+      relevant_status_reports.any? { |report| report.missing_checks.present? }
+    end
+
+    def check_state
```
Comment (vpereira) [inline src/api/app/models/staging/staging_project.rb]: Why not:

```
def check_state
  status_checks = Status::Check.where(status_reports_id: relevant_status_reports)
  if missing_checks? || status_checks.pending.exists?
    :testing
  elsif status_checks.failed.exists?
    :failed
  else
    :acceptable
  end
end

```

File: src/api/app/models/staging/staging_project.rb [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+
+      overall_state
+    end
+
+    def build_state
+      set_buildinfo
+
+      return :building if building_repositories.present?
+      return :failed if broken_packages.present?
+
```
[the remark this answers is not in the document]
Comment (coolo) [reply src/api/app/models/staging/staging_project.rb]: guess it's a matter of taste, but I find the "die, die" more readable

File: src/api/app/models/staging/staging_project.rb (hunk 18 above)
Comment (bgeuken) [reply src/api/app/models/staging/staging_project.rb]: We had to pick one. And since there is a slight preference within the team to use guard clauses, we picked that one.

File: src/api/app/models/staging_project.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/models/staging_project.rb]: I know how `inverse_of` works :unamused: I meant that it was in the wrong place. As I told @bgeuken, it needs to be in the other relation. There is no rails magic, it is just needed if you don't use the model name as the relationship name.

File: src/api/app/models/staging/staging_project.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (vpereira) [reply src/api/app/models/staging/staging_project.rb]: For a 4 lines of code? 🤔🤔

File: src/api/app/models/staging/staging_project.rb [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+    def missing_reviews
+      return @missing_reviews if @missing_reviews
+
+      @missing_reviews = []
+      attribs = [:by_group, :by_user, :by_project, :by_package]
+
+      (requests_to_review + staged_requests).uniq.each do |request|
+        request.reviews.each do |review|
+          next if review.state == :accepted || review.by_project == name
```
Comment (Ana06) [inline src/api/app/models/staging/staging_project.rb]: I still think this should be moved to a `where`. It should be way more efficient [although I would agree to this being done in another PR to get this merged]

File: src/api/app/models/staging/staging_project.rb [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+          # FIXME: this loop (and the inner if) would not be needed
+          # if every review only has one valid by_xxx.
+          # I'm keeping it to mimic the python implementation.
```
Comment (Ana06) [inline src/api/app/models/staging/staging_project.rb]: @bgeuken 

> That method is just copied over from the old staging project code. We kept it for reference.

I though we don't want just to copy code over but to adapt it. I would write this properly, with a `where` instead of keeping it coherent with some Python code we don't even know where it is and if still keep looking the same. [as said before I would agree to this being done in another PR to get this merged]

File: src/api/app/models/staging/staging_project.rb (hunk 18 above)
Comment (Ana06) [reply src/api/app/models/staging/staging_project.rb]: @vpereira option is repeating some code `status_checks.failed.exists?`. I also find the other (original) more readable.

File: src/api/spec/models/staging/staging_project_spec.rb [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+
+  let!(:submit_request) { create(:bs_request_with_submit_action, request_attributes.merge(staging_project: staging_project)) }
+
+  before do
+    allow(Backend::Api::Published).to receive(:build_id).with(staging_project.name, repository.name).and_return('1234')
+    status_report.update(uuid: '1234')
+  end
+
+  describe '#staging_identifier' do
+    let(:staging_project) { create(:staging_project, name: 'openSUSE_41:Staging:myStagingProject') }
```
Comment (Ana06) [inline src/api/spec/models/staging/staging_project_spec.rb]: why do you need another `staging_project`? there is one here already: https://github.com/openSUSE/open-build-service/pull/6224/files#diff-5e20b29a349feec5580923e52507b03eR7 :confused:

File: src/api/app/models/staging/staging_project.rb (hunk 20 above)
Comment (bgeuken) [reply src/api/app/models/staging/staging_project.rb]: As mentioned before. This method is going to be refactored and updated (droping requests_to_review) soon:-)

File: src/api/spec/models/staging/staging_project_spec.rb (hunk 22 above)
Comment (bgeuken) [reply src/api/spec/models/staging/staging_project_spec.rb]: Right, I'll update it. Reason for that is that we have 5 or 6 branches that touch the same code. So every time our feature branch updates, I have to update the code here... Time to get this merged;-)

File: src/api/app/models/staging/staging_project.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (dmarcoux) [reply src/api/app/models/staging/staging_project.rb]: @vpereira: This is a common pattern -> https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

File: src/api/app/models/staging/staging_project.rb (hunk 18 above)
Comment (dmarcoux) [reply src/api/app/models/staging/staging_project.rb]: @vpereira: Same as above, this is a common pattern -> https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html

Comment (vpereira) [reply src/api/app/models/staging/staging_project.rb]: yes guard clauses are really common, however i dont see a nested conditional here.. interesting that codeclimate do not complain. SonarQube uses the ratio between number of lines x number of returns in the cyclomatic complexity calculation..

File: src/api/app/models/staging/staging_project.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (vpereira) [reply src/api/app/models/staging/staging_project.rb]: which nested condition are you replacing here @dmarcoux?

File: src/api/app/models/staging/staging_project.rb (hunk 18 above)
Comment (vpereira) [reply src/api/app/models/staging/staging_project.rb]: > @vpereira option is repeating some code `status_checks.failed.exists?`. I also find the other (original) more readable.

@Ana06 where is the repeated code? The conditions are the same, just instead of return, return, if, elsif, else..

File: src/api/app/models/staging/staging_project.rb [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+      set_buildinfo unless @building_repositories
+      @building_repositories
+    end
+
+    def broken_packages
+      set_buildinfo unless @broken_packages
+      @broken_packages
+    end
+
+    def missing_reviews
```
Comment (vpereira) [inline src/api/app/models/staging/staging_project.rb]: no test code for missing_reviews, which we could use to help in the next refactoring iteration @bgeuken ?

File: src/api/app/models/staging/staging_project.rb [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+    end
+
+    def check_state
+      status_checks = Status::Check.where(status_reports_id: relevant_status_reports)
+      return :testing if missing_checks? || status_checks.pending.exists?
+      return :failed if status_checks.failed.exists?
+      return :acceptable
+    end
+
+    def set_buildinfo
```
Comment (vpereira) [inline src/api/app/models/staging/staging_project.rb]: @bgeuken @dmarcoux could we test little bit more this 50+ lines method? https://coveralls.io/builds/20080518/source?filename=app/models/staging/staging_project.rb#L36

File: src/api/app/models/staging/staging_project.rb (hunk 18 above)
Comment (vpereira) [reply src/api/app/models/staging/staging_project.rb]: @coolo its plain taste, yes thats the only valid argument here. No nested conditional, the code - in a 4 lines of code method - with if/elseif/else isnt less readable.. is just taste. Thats why I`m just  commenting and not asking for change.

File: src/api/app/models/staging/staging_project.rb (hunk 23 above)
Comment (bgeuken) [reply src/api/app/models/staging/staging_project.rb]: Right. As stated before this method is going to change. I was just waiting for some feedback. So I paused writing the tests for this part.

File: src/api/app/models/staging/staging_project.rb (hunk 24 above)
Comment (bgeuken) [reply src/api/app/models/staging/staging_project.rb]: And this one would require some more effort, and thinking, to write a meaningful test. That's because it depends so much on the backend and the xml it returns.

Keep in mind that PR is not going to master, but to a WIP branch:wink:

File: src/api/app/models/staging/staging_project.rb (hunk 18 above)
Comment (bgeuken) [reply src/api/app/models/staging/staging_project.rb]: So you agree that you have no argument for not using guard clauses and this discussion is not going anywhere?:-)

Comment (vpereira) [reply src/api/app/models/staging/staging_project.rb]: a suggestion don't have to go anywhere. Don't take it personal, to write perl looking ruby is quite ok here ;)

Comment (vpereira) [reply src/api/app/models/staging/staging_project.rb]: btw the codecoverage handles as bad smell 4 or more return-statements per method.
